### PR TITLE
DPE-4235, DPE-3262, DPE-4315 Common UX for replication

### LINKED ISSUE #445: async re-relation ofter switchover breaks charm
## Steps to reproduce

1. Deploy A (3 units)
2. Deploy B (1 unit)
3. A=>async=>B ( I also have continuous writes.. but should change nothing).
4. Make B primary
5. Remove relation A=>async=>B (A will be Standalone).
6. Create relation B=>A 

## Expected behavior

Relation should render cluster set functional

## Actual behavior

Both clusters are broken

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

* See issue #445
* async relation secret was not being removed on relation broken
* non-leader hooks were being run by the leader
* Without a common UX with PG
* Password is not propagated across clusters when rotated

## Solution

* Block on relation from the "wrong" side, i.e., a blocked cluster cannot be the `async-primary` when creating the async relation.
* removes relation secret on relation broken
* implement password rotation from offer side only
* add assumes juju >=3.4.3 for full replication support

Fixes #445

### CODE REVIEW

**Review by taurus-forever — APPROVED** [review]
Nice! Will test this tomorrow in Madrid! See you there!

File: lib/charms/mysql/v0/async_replication.py [hunk 1]
```diff
@@ -221,17 +224,21 @@ def on_async_relation_broken(self, event):  # noqa: C901
...
-                )
-                # reset the cluster node count flag
-                del self._charm.app_peer_data["units-added-to-cluster"]
-                # set flag to persist removed from cluster-set state
-                self._charm.app_peer_data["removed-from-cluster-set"] = "true"
+            if not self._charm.unit.is_leader():
+                # delay non leader to avoid `update_status` running before
+                # leader updates app peer data
+                sleep(10)
+                return
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/async_replication.py:231]: Permanent. The delay sole purpose is to make the status setting for non-leader units happen after the leader, the effect being proper unit status being set faster, instead of having to wait for the next hook (worse case, update-status).
tl;dr; it's just cosmetics

**Review by taurus-forever — APPROVED** [review]
Excellent!

File: lib/charms/mysql/v0/async_replication.py [hunk 2]
```diff
@@ -481,53 +574,67 @@ def _on_primary_relation_changed(self, event):
...
+            # remove relation secret by id
+            if secret_id := event.relation.data[self.model.app].get("secret-id"):
+                logger.debug("Removing replication secret")
+                secret = self._charm.model.get_secret(id=secret_id)
+                secret.remove_all_revisions()
+            else:
+                logger.debug("Secret-id not set, skipping removal")
+
+    def _on_secret_change(self, event: SecretChangedEvent):
+        """Propagates the secret being changed."""
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/async_replication.py:589]: yes.
It should but that add extra complexity since we always deploy an app with it own secrets, so unifying old need to delete one, and recreating on relation broken.
giving the troubled history with secrets and CMR, I rather do some duplication that works for now.

File: metadata.yaml [hunk 3]
```diff
@@ -66,3 +67,10 @@ storage:
 
 assumes:
   - juju
+  - any-of:
+      - all-of:
+          - juju >= 2.9.44
```
[the remark this answers is not in the document]
Comment (paulomach) [reply metadata.yaml:72]: replication will not work for juju <3
so this is enough for other features we have already besides replication

File: lib/charms/mysql/v0/async_replication.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/async_replication.py:589]: It can't. Juju design it thinking in a saas model, where a service is provider. In practice the relation offer side is the only one that can manipulate the secret granted to the relation. The consumer side can neither write nor grant a secret to the relation.
